pkg/resource/deploy/source_eval.go

```diff
@@ -773,22 +773,26 @@ func (rm *resmon) RegisterResource(ctx context.Context,
 	label := fmt.Sprintf("ResourceMonitor.RegisterResource(%s,%s)", t, name)
 
 	var providerRef providers.Reference
+	var providerRefs map[string]string
+
 	if custom && !providers.IsProviderType(t) || remote {
-		provider := req.GetProvider()
-		if provider == "" {
-			providerReq, err := parseProviderRequest(t.Package(), req.GetVersion())
-			if err != nil {
-				return nil, err
-			}
-			providerRef, err = rm.defaultProviders.getDefaultProviderRef(providerReq)
+		providerReq, err := parseProviderRequest(t.Package(), req.GetVersion())
+		if err != nil {
+			return nil, err
+		}
+
+		providerRef, err = getProviderReference(rm.defaultProviders, providerReq, req.GetProvider())
+		if err != nil {
+			return nil, err
+		}
+
+		providerRefs = make(map[string]string)
```

Comment:
Nit: Do we want to pass the length of `req.GetProviders()` here? Or not allocate at all if it's empty?

Reply:
Sure, might as well.